### DISCUSSION TITLE
Enable users to control the title bar

### DISCUSSION THREAD

**Discussion by charles-esterbrook** [zim-desktop-wiki/zim-desktop-wiki#1660] [Ideas]
## Title bar template preference

Zim would benefit from a new Interface preference that uses a Python templated string to let the user control the title bar. 

Fields could include:

- page name ("ZimIsGreat")
- page location ("NoteTaking:Applications")
- page path ("NoteTaking:Applications:ZimIsGreat")
- notebook name ("My Life")
- notebook path ("/home/me/Notes/MyLife")

Examples:

- `{notebook name} - Zim`
  - "My Life - Zim"
  - Zim 0.6x style
- `{page path} - {notebook name}`
  - "NoteTaking:Applications:ZimIsGreat - My Life"
  - Zim 0.74 style
- `{notebook name} - {page path}`
  - "My Life - NoteTaking:Applications:ZimIsGreat"
  - for #1339
- `Zim: {notebook name} in {notebook path}`
  - "Zim: My Life in /home/me/Notes/MyLife"
- `Page {page name} in {notebook name} - Zim`
  - "Page ZimIsGreat in My Life"
- etc.

Python's .format() method would be called on the template as in `titleBarTemplateStr.format(**fieldsDict)`, probably with `try:except KeyError:` protection.

## Motivation

When moving form Zim 0.6x to Zim 0.7x, I can no longer efficiently use the Windows 10 taskbar window previews to choose a notebook, due to changes in the title bar. Note that the window previews are shown when left clicking the Zim app icon in the taskbar. Right clicking shows a basic menu ("Zim Desktop Wiki", "Unpin from taskbar", "Close all windows").

You can see the taskbar-triggered window previews in the screenshot at this unrelated superuser issue:
https://superuser.com/questions/1136612/how-to-refresh-windows-10-taskbar-thumbnail-previews

Note that all the title bars are clipped with "...". You can only read the beginning of each title bar. This is key.

Suppose you have multiple Zim notebooks open and their names are:

- Artificial Intelligence
- TechNotes

Then the title bars in Zim 0.6x are:

- Artificial Intelligence - Zim
- TechNotes - Zim

And the Windows 10 window previews will show:

- Artificial Intel...
- TechNotes...

Which is easy to use (meaning choose from) because these are the names of the notebook, or at least their beginnings.

However, in Zim 0.7x, the title bars will start with the _current page name_ which changes quite frequently. Note only can the constant change of the beginning of the title bar make it more difficult to pick the notebook from a list, but two notebooks may have the identical name among the windows previews. Suppose you are looking at the page "Applications" in both notebooks. Then their window title bars are:

- Applications - Artificial Intelligence
- Applications - TechNotes

But the window list shows:

- Applications ...
- Applications ...

... that's _my_ situation. In #1339, another user wants to put the full path of the page in the title bar - after the notebook name. I personally don't want the page name let alone its full path, but I understand that people have different needs which may also vary by window manager. My needs could even change if I move from Windows to Mac or Linux.

**Comment by jaap-karssenberg** [comment]
Open to have this configureable but would like to have it as a "hidden"
preference - or in some kind of "advanced" section of the preferenes
dialog.

**Comment by charles-esterbrook** [comment]
Even after googling, I'm not clear on why a proposed preference is moved from "issues" to "discussions". But I'm new to github discussions.

On a related note, I would call the title bars in 0.7x a regression from 0.69 where I could quickly see the notebook name. Now I see "FolderA:FolderB:..."

@jaap-karssenberg Can you share how you perceive discussions vs. issues?

**Comment by jaap-karssenberg** [comment]
Trying to make a change where the issues are reserved for clear bugs &
concrete work items. Proposals should go to discussions first and turn into
an issue ticket when someone is actually working on it - or even better,
into a pull request. Still testing this out but need to get influx of new
issue tickets under control.

**Comment by GHSoCoder** [comment]
Yep, same wish - Windows guy! Doing a lot of automation with [AutoHotkey](https://www.autohotkey.com/) (AHK). The text of the title bar is one of the most important parts in AHK - mainly for activating (getting focus of) the window (app). But I could do other crazy stuff with its content.
So I just miss the old text part " - Zim" - any fast solution without recompiling?

**Comment by charles-esterbrook** [comment]
@GHSoCoder The change in title bars from 0.69 is still an issue/bug/regression for me. But this "github issue" has been changed to a "github discussion" so I don't think this is happening (soon). I don't know of any hacks to address it, but maybe @jaap-karssenberg does.

**Comment by GHSoCoder** [comment]
@charles-Esterbrook Therefore I have just modified my AHK script so that the Zim window is activated via its process name Zim.exe now. That solution is not so flexible compared to a configuarable title bar but it works as a short-term solution!
BTW, for the Visual Studio users here (and as inspiration...), take a look at [Customize Visual Studio Window Title - Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=mayerwin.RenameVisualStudioWindowTitle).

**Comment by jaap-karssenberg** [comment]
> But this "github issue" has been changed to a "github discussion" so I
don't think this is happening (soon)

Indeed this is not high on my own prio list, but I'm always open to pull
requests.

-- Jaap

**Comment by joedarwilkat** [comment]
This is an issue for me too.  I am a long term user of Zim and generally have tow or more notes open with icons in the panel at the bottom of the screen.  In the latest version, I can no longer distinguish which notebook each icon refers to.  I would like the Window Title to revert to just the Notebook Name.  Is there a fix for this?

**Comment by charles-esterbrook** [comment]
This is still painful and also a reversion in usability. Should be an issue rather than discussion.

**Comment by introt** [comment]
You've been heard. Let me (try to) show you how it's done!

...as in how the Discussion -> Issue -> PR workflow works. #2017 shows fellow devs that I've started working on this, so we don't end up doing duplicates. Sure, could've skipped straight to PR at this point in this particular instance, but if I didn't go the monkey patching route it would've been very useful as a place to co-ordinate related API rework on the level of "concrete work items", while keeping the other discussion (workarounds etc) here, out of the way.

**Reply by introt** [reply]
To clarify from a dev's perspective, ideally "Issues" are for what _**is being** worked on_ right now (blueprinted roadmap items, reproducible bugs, etc - these are a _**sub**set_ of user's issues, and ideally there's only a handful open at a time), while "Discussions" are for stuff that _**could be** worked on_ (as soon as anyone finds the time, or in the future). 

It's hard to prioritize given the backlog of almost 500 open issues - is title bar control more important than say, markdown support? - but we try our best to balance our time. Please keep in mind that this is free, open source software made by volunteers - anyone can work on anything they choose at any time, regardless of whether there's an open issue or not.

For more information about issues and discussions, refer to #1610. Thanks for your understanding.

**Comment by charles-esterbrook** [comment]
Lack of markdown support is not a reversion when moving from Zim 0.6x to Zim 0.7x. It's also a much more complicated feature. Not a good comparison.

**Reply by dehesselle** [reply]
Why are you splitting hairs here? You know exactly how it was meant what introt said. And since I don't see you writing the code for this, do you think it is wise to annoy the people who do?

**Reply by introt** [reply]
My apologies @charles-esterbrook, my comments written in good faith, though I do see how they could be interpreted as something else.

**Reply by charles-esterbrook** [reply]
No worries, @introt. Sorry if I sounded snippy!

**Reply by introt** [reply]
It's all good. Let me know what you think of Window Title Editor when you get the chance! It was a fun one to implement - there aren't too many cases where monkey patching makes this much sense to do in production code - made me feel a bit like an actual hacker in the problem solving sense :)

**Comment by introt** [comment]
@GHSoCoder @charles-esterbrook @joedarwilkat

I've opened PR #2020, please check it out if you can and let me know if there's anything you'd like to see change(d), thanks!

No need to mess with `git`to try it out either, just put the file into your plugins folder:

https://raw.githubusercontent.com/introt/zim-desktop-wiki/windowtitleeditor/zim/plugins/windowtitleeditor.py

**Reply by charles-esterbrook** [reply]
Works perfect. I changed the template from `Zim: $page - $notebook ($source)` to `$notebook Zim ($source)` and now I can instantly see which window is which notebook! Thanks, @introt!

**Reply by charles-esterbrook** [reply]
In case anyone needs this, the steps:

1. Menu > Edit > Preferences
1. Plugins tab
1. Click "Open plugins folder" button
1. Put the file (windowtitleeditor.py) in there
1. I restarted Zim; may not be necessary
1. Back to Plugins, check off "Window Title Editor"
1. Click "Configure"
1. Edit the template to start with $notebook
1. Enjoy the productivity

**Reply by introt** [reply]
You're welcome!

> from `Zim: $page - $notebook ($source)`

Ah, I knew I forgot something - I made the default a bit tacky so I'd notice whether loading/unloading worked ok.

For step 8, here's the manual page with a list of available substitutions: https://raw.githubusercontent.com/zim-desktop-wiki/zim-desktop-wiki/02c1bbd4a6efde57eadbdecb182fe5d6952aa577/data/manual/Plugins/WindowTitleEditor.txt

Have a good one!

**Reply by tymmek** [reply]
@introt Where can I add that **manual page** ?

In the Plugins window, hitting "More" looks for the help file under
/usr/share/zim/manual/Plugins/…

Is there a friendlier way of adding it within the home directory just like the plugin itself?

By the way, amazing work -- just what I was looking for -- THANK YOU.

**Reply by introt** [reply]
@tymmek Thanks, you're very welcome! It was a fun one to code, too :)

As far as I'm aware there isn't currently a way to add pages to the manual besides editing the notebook included in the installation. I've described one possible solution for this [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/1936#discussioncomment-2839581).

(Note to self: there seems to be a typo in the 4th example.)

**Comment by joedarwilkat** [comment]
Thank you introt and charles-esterbrook.  This worked perfectly for me once I found my plugins folder at:
/home/<user>/.var/app/org.zim_wiki.Zim/data/zim/plugins
If only I was a programmer :-(

**Reply by introt** [reply]
You're very welcome (to turn that frown upside down, it's never too late to learn programming and there's a plethora of free courses & other material available online ;)!

**Comment by mtill** [comment]
love the idea of being able to control the title bar.. cant wait to have that in one of the next releases! BR

**Reply by mtill** [reply]
btw: is there a timeline for the next release ?

**Reply by introt** [reply]
> cant wait to have that in one of the next releases!

No need to wait - [the plugin](https://raw.githubusercontent.com/introt/zim-desktop-wiki/windowtitleeditor/zim/plugins/windowtitleeditor.py) supports the current release!

> btw: is there a timeline for the next release ?

Not that I know of.

**Reply by mtill** [reply]
thanks for sharing! works great!
one important thing however is to disable the ToolBar plugin as otherwise it will not work correctly
BR

**Reply by introt** [reply]
Hi @mtill, could you please elaborate on that? I didn't come across any issues with or without "Show controls in the window decoration" option enabled.

**Reply by mtill** [reply]
Hi @introt , it did not work in that case on a windows machine - will  come back with some more details soon

**Reply by mtill** [reply]
@introt, happens when I enable Tool Bar plugin, (I disabled all options). After restarting ZIM on windows, the window title editor plugin stops working. BR

**Reply by charles-esterbrook** [reply]
I couldn't recreate this. I'm on Windows 10 and Zim 0.74.1. When I enable the Tool Bar, the "Windows Title Editor" continues to work including after restarting Zim. I don't know what you mean by "(I disable all options)." -- the toolbar has choices for placement etc and the title editor has the text for the title bar. Feel free to elaborate.

I grabbed a fresh copy of the plugin (mine was from Apr 24), but the only changes were a string literal and a comment. Tried to recreate the problem again, but still couldn't.

There must be some other difference between our set-ups.

Check the menu > Tools > View debug log, to see if there are any error messages.

**Reply by mtill** [reply]
thank you very much for your reply!
I checked the debug files and as far as I can see there this problem seems to be related to the toolbar plugin itself. the toolbar plugin seems to crash and as a result, the windowtitle plugin is not initialized by zim. thus i think the problem has nothing to do with windowtitle plugin.

**Reply by introt** [reply]
> the toolbar plugin seems to crash 

That sounds like a bug, please fill out a report form, thanks!

**Comment by FabianF95** [comment]
I love it. Will improve my performance significantly as it stops the frustrating "Damn, wrong notebook again" ;-)
